### PR TITLE
feat(dashboard): add support for additional metrics in overview

### PR DESCRIPTION
**Pull request by onehassan**
### **User description**
resolves https://github.com/nhost/nhost/issues/3017

___

### **PR Type**
Enhancement, Other

___

### **Description**
- Introduced new metrics in the dashboard overview, including monthly and daily active users, total users, and storage.
- Implemented a new GraphQL query `GetUserProjectMetrics` to fetch user-related metrics.
- Updated the `OverviewMetrics` component to display the newly added metrics.
- Enhanced the GraphQL schema with new fields and queries to support the additional metrics.
- Added a changeset to document the new feature in the dashboard.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>OverviewMetrics.tsx</strong><dd><code>Add support for additional user and storage metrics in overview</code></dd></summary>
<hr>

dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx

<li>Added new metrics for monthly and daily active users, total users, and <br>storage.<br> <li> Integrated <code>useGetUserProjectMetricsQuery</code> for fetching user-related <br>metrics.<br> <li> Updated metrics card elements to display new metrics.<br> <li> Removed redundant data checks and improved error handling.<br>

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/3052/files#diff-de881837e53f594075bb725282b02e92c2cb281f8f6a438fdbaa2e3254907fd1">+90/-17</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>graphql.ts</strong><dd><code>Update GraphQL types and queries for user metrics</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dashboard/src/utils/__generated__/graphql.ts

<li>Added new GraphQL query types for user project metrics.<br> <li> Introduced <code>automaticDeploys</code> field in the <code>Apps</code> type.<br> <li> Updated generated types to include new fields and queries.<br>

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/3052/files#diff-fbd5db84b560b1c91675004448c6c7fa0dcbfb28b9eb05d53b03e6cb7b83ebac">+87/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>getUserProjectMetrics.gql</strong><dd><code>Add GraphQL query for user project metrics</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dashboard/src/gql/organizations/getUserProjectMetrics.gql

- Created new GraphQL query for fetching user project metrics.

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/3052/files#diff-902523302d8b32d218ef665a252dec5b9cbcf5fbab0cbb32845c441b01eaa28e">+28/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>khaki-pets-argue.md</strong><dd><code>Add changeset for dashboard metrics feature</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.changeset/khaki-pets-argue.md

- Added changeset for new feature in dashboard.

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/3052/files#diff-51eb36ca77dae90a865c185e0d528fcaa4b836f3a0469550513f68003bf11c9a">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx [hunk 1]
```diff
@@ -1,54 +1,127 @@
...
-        'Total amount of requests your services have received excluding functions',
-      value: prettifyNumber(data?.totalRequests?.value || 0, {
-        numberOfDecimals: data?.totalRequests?.value > 1000 ? 2 : 0,
+      tooltip: 'Total service requests (excluding functions)',
+      value: prettifyNumber(totalRequests || 0, {
+        numberOfDecimals: totalRequests > 1000 ? 2 : 0,
       }),
     },
     {
       label: 'Function Invocations',
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:93]: let's remove function invocations so have 9 in total

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx [hunk 2]
```diff
@@ -1,54 +1,127 @@
...
+      value: prettifyNumber(monthlyActiveUsers),
+    },
+    {
+      label: 'Daily Active Users',
+      tooltip: 'Unique users active today',
+      value: prettifyNumber(dailyActiveUsers),
+    },
+    {
+      label: 'All Users',
+      tooltip: 'Total unique users',
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:77]:
```suggestion
      tooltip: 'Total registered users',
```

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx [hunk 3]
```diff
@@ -1,54 +1,127 @@
...
+      value: prettifyNumber(allUsers),
+    },
+    {
+      label: 'Storage',
+      tooltip: 'Total size of stored files',
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:82]:
```suggestion
      tooltip: 'Total size of stored files in the storage service',
```

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx [hunk 4]
```diff
@@ -1,54 +1,127 @@
...
+      value: prettifySize(totalStorage || 0),
     },
     {
       label: 'Total Requests',
-      tooltip:
-        'Total amount of requests your services have received excluding functions',
-      value: prettifyNumber(data?.totalRequests?.value || 0, {
-        numberOfDecimals: data?.totalRequests?.value > 1000 ? 2 : 0,
+      tooltip: 'Total service requests (excluding functions)',
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:87]: which timescale? maybe we can add function invocations here too

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx [hunk 5]
```diff
@@ -1,54 +1,127 @@
...
+      value: prettifyNumber(functionInvocations || 0, {
         numberOfDecimals: 0,
       }),
     },
     {
-      label: 'Logs',
-      tooltip: 'Amount of logs stored',
-      value: prettifySize(data?.logsVolume?.value || 0),
+      label: 'Function Duration',
+      tooltip: 'Total function execution time',
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:101]:
```suggestion
      tooltip: 'Total function execution time this month',
```

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx [hunk 6]
```diff
@@ -1,54 +1,127 @@
...
+      value: prettifyNumber(functionsDuration),
+    },
+    {
+      label: 'Postgres Volume Usage',
+      tooltip: 'Used storage in the Postgres database',
+      value: prettifySize(postgresVolumeUsage),
+    },
+    {
+      label: 'Egress',
+      tooltip: 'Total outgoing data transfer',
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:111]: timescale?

Comment (onehassan) [reply dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:111]: From start of the month until today

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx (hunk 4 above)
Comment (onehassan) [reply dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:87]: From start of the month until today

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx (hunk 1 above)
Comment (onehassan) [reply dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:93]: 👍

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx (hunk 6 above)
Comment (dbarrosop) [reply dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:117]:
```suggestion
      tooltip: 'Total outgoing data transfer this month so far',
```

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx [hunk 7]
```diff
@@ -1,54 +1,132 @@
...
       }),
     },
     {
-      label: 'Function Invocations',
-      tooltip: 'Number of times your functions have been called',
-      value: prettifyNumber(data?.functionInvocations?.value || 0, {
-        numberOfDecimals: 0,
-      }),
+      label: 'Functions Duration',
+      tooltip: 'Total Functions execution time this month',
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:107]:
```suggestion
      tooltip: 'Total Functions execution this month so far',
```

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx (hunk 4 above)
Comment (dbarrosop) [reply dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:100]:
```suggestion
      tooltip: 'Total service requests this month so far (excluding functions)',
```

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx [hunk 8]
```diff
@@ -1,54 +1,132 @@
...
     },
     {
       label: 'Total Requests',
-      tooltip:
-        'Total amount of requests your services have received excluding functions',
-      value: prettifyNumber(data?.totalRequests?.value || 0, {
-        numberOfDecimals: data?.totalRequests?.value > 1000 ? 2 : 0,
+      tooltip: 'Total service requests (excluding functions)',
+      value: prettifyNumber(totalRequests || 0, {
+        numberOfDecimals: totalRequests > 1000 ? 2 : 0,
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:102]: I am a bit dumb and I am failing to understand how you are getting this number. I only see the query that limits this to 5 minutes

**Comment by dbarrosop** [comment]
Just two last comments I think:

1. regarding totalrequests, see the inline comment
2. Re the order, let's change it so it is:
  1. DAU, MAU, Total users
  2. RPS, Total Requests, Egress
  3. Functions Duration, Storage, Postgres volume

File: dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx (hunk 8 above)
Comment (onehassan) [reply dashboard/src/features/orgs/projects/overview/components/OverviewMetrics/OverviewMetrics.tsx:102]: I'm doing two queries:

1. This query gets total requests in the last 5 minutes which I'm polling once every 5mns

https://github.com/nhost/nhost/pull/3052/files#diff-de881837e53f594075bb725282b02e92c2cb281f8f6a438fdbaa2e3254907fd1R47

```ts
const {
  data: {
    totalRequests: { value: totalRequestsInLastFiveMinutes = 0 } = {},
  } = {},
} = useGetProjectRequestsMetricQuery({
  variables: {
    appId: project.id,
    from: formatISO(subMinutes(new Date(), 6)), // 6 mns earlier
    to: formatISO(subMinutes(new Date(), 1)), // 1 mn earlier
  },
  skip: !project,
  pollInterval: 1000 * 60 * 5, // Poll every 5 minutes
});
```

2. This other one pulls totalRequests and other metrics only once for the current month

https://github.com/nhost/nhost/pull/3052/files#diff-de881837e53f594075bb725282b02e92c2cb281f8f6a438fdbaa2e3254907fd1R62

```ts
const {
  data: {
    functionsDuration: { value: functionsDuration = 0 } = {},
    totalRequests: { value: totalRequests = 0 } = {},
    postgresVolumeUsage: { value: postgresVolumeUsage = 0 } = {},
    egressVolume: { value: egressVolume = 0 } = {},
  } = {},
  loading,
  error,
} = useGetProjectMetricsQuery({
  variables: {
    appId: project.id,
    subdomain: project?.subdomain,
    from: new Date(now.getFullYear(), now.getMonth(), 1),
  },
  skip: !project,
});
```

**Pull request merged**